app/models/user.rb

```diff
@@ -400,12 +400,20 @@ def taxonomy_ids
   end
 
   def visible_environments
-    Environment.unscoped
-               .authorized(:view_environemnts)
-               .joins(:taxable_taxonomies)
-               .where('taxable_taxonomies.taxonomy_id' => taxonomy_ids[:organizations] + taxonomy_ids[:locations])
-               .distinct
-               .pluck(:name)
+    authorized_scope = Environment.unscoped.authorized(:view_environments)
+    if Taxonomy.locations_enabled || Taxonomy.organizations_enabled
+      authorized_scope = authorized_scope.
+        joins(:taxable_taxonomies)
+        .where('taxable_taxonomies.taxonomy_id' => taxonomy_ids[:organizations] + taxonomy_ids[:locations])
+    end
+    result = authorized_scope.distinct.pluck(:name)
+    if User.current.admin
+      # Admin users can also see Environments that do not have any organization or location, even when
+      # organizations and locations are enabled.
+      untaxed_environments = Environment.unscoped.select { |e| e.locations == [] || e.organizations == [] }.map(&:name)
```

Comment:
This would load all environments to memory, I wonder if it's possible to achieve the same result in another way? if we load everything to memory might as well do all the filtering in memory